### ISSUE TITLE
AWS Credential Refresh Issue

### ISSUE THREAD

**Issue by kjdelisle** [tulios/kafkajs#423]
Hello again!

Normally for something like this, I would've written a pull request first, but I wanted to coordinate on this to make sure I'm not hacking through existing patterns and whatnot.

The problem I'm currently seeing with the new AWS authentication is that it won't work for session-based credentials (which expire and require refresh). Originally, I thought that the consuming code would need to be responsible for handling these refreshes, but after some investigative work I've realized that Kafka.js is caching these credentials at various layers.

Even after writing a custom getter and injecting it into the options, I've noticed that it never gets called more than once, which means its output is being cloned on the first call.

Is there an established methodology for this that I'm not seeing, or would there need to be some refactoring involved to hoist those credentials out into some sort of provider that can be updated on-the-fly?

As of right now, the only alternative I can see is disposing of the Client on each call -- there aren't any errors that my consuming code can capture directly (please let me know if this is wrong), which means that I can't trigger a new Client creation on authentication failure.

Given the expense of creating all of the broker connections again, that's not something I think we'd want to do.

**Comment by tulios** [comment]
Yes, Kafka updated the spec to support a refresh cycle. I have this in a list somewhere; I can get back to you soon.
